src/lightning/pytorch/loops/fit_loop.py

```diff
@@ -409,9 +414,34 @@ def teardown(self) -> None:
             self._data_fetcher = None
         self.epoch_loop.teardown()
 
+    @override
+    def on_save_checkpoint(self) -> Dict:
+        state_dict = super().on_save_checkpoint()
+        loaders = self._combined_loader.flattened if self._combined_loader is not None else []
+        loader_states = [loader.state_dict() for loader in loaders if isinstance(loader, _Stateful)]
+        if loader_states:
+            state_dict["combined_loader"] = loader_states
+        return state_dict
+
+    @override
+    def on_load_checkpoint(self, state_dict: Dict) -> None:
+        self._combined_loader_states = state_dict.get("combined_loader")
+        super().on_load_checkpoint(state_dict)
+
     def _should_accumulate(self) -> bool:
         """Whether the gradients should be accumulated."""
         return self.epoch_loop._should_accumulate()
 
     def _iteration_based_training(self) -> bool:
         return self.trainer.max_steps != -1
+
+    def _restore_combined_loader_state(self) -> None:
+        if not self._combined_loader_states:
+            return
+
+        loaders = self._combined_loader.flattened if self._combined_loader is not None else []
+        stateful_loaders = [loader for loader in loaders if isinstance(loader, _Stateful)]
+        for loader, state_dict in zip(stateful_loaders, self._combined_loader_states):
```

Comment:
zip will allow that stateful_loaders and combined_loader_states have different lengths. It's probably safer to raise an error if they have different lengths

Reply:
Good idea, I added the error checking